### PR TITLE
Feature/bitbucket integration

### PR DESCRIPTION
**Pull request by nikoleta-vukajlovic**
Fetching all workspaces for bitbucket, all repositories for workspace, all namspaces for gitlab, all repositories for namspace, all branches for repository for github and bitbucket

### CODE REVIEW

File: src/components/GitlabApp/service.ts [hunk 1]
```diff
@@ -97,13 +99,78 @@ const GitlabAppService: IGitlabAppTokenService = {
...
+
+    return repositories;
+  },
+  async getNamespaces(id: string): Promise<any> {
+    let getUserToken = await GitlabAppService.findByUserId(Types.ObjectId(id));
+    if (new Date() > getUserToken.expiresAt) {
+      getUserToken = await GitlabAppService.refreshToken(Types.ObjectId(id));
+    }
+
     const projectInfo = await axios.get(
```
Comment (rekpero) [inline src/components/GitlabApp/service.ts:123]: rename the variable to `namespaces`

File: src/components/GitlabApp/service.ts [hunk 2]
```diff
@@ -97,13 +99,78 @@ const GitlabAppService: IGitlabAppTokenService = {
...
+      );
+      for (const repo of response) {
+        repositories.push(repo);
+      }
+    }
+
+    return repositories;
+  },
+  async getNamespaces(id: string): Promise<any> {
+    let getUserToken = await GitlabAppService.findByUserId(Types.ObjectId(id));
```
Comment (rekpero) [inline src/components/GitlabApp/service.ts:118]: rename it to `gitUserToken`

File: src/components/GitlabApp/service.ts [hunk 3]
```diff
@@ -97,13 +99,78 @@ const GitlabAppService: IGitlabAppTokenService = {
...
+      }
+    );
+
+    return projectInfo;
+  },
+  async getNamespaceRepositories(
+    id: string,
+    namespaceId: string
+  ): Promise<any> {
+    let getUserToken = await GitlabAppService.findByUserId(Types.ObjectId(id));
```
Comment (rekpero) [inline src/components/GitlabApp/service.ts:136]: same here, rename it to `gitUserToken`

File: src/components/GitlabApp/service.ts [hunk 4]
```diff
@@ -97,13 +99,78 @@ const GitlabAppService: IGitlabAppTokenService = {
...
+      projectInfo = await axios.get(
+        `https://gitlab.com/api/v4/groups/${namespaceId}/projects`,
+        { headers: { authorization: `bearer ${getUserToken.accessToken}` } }
+      );
+    }
 
     return projectInfo.data;
   },
+  async getBranches(id: string, branchesQuery: any): Promise<any> {
+    let getUserToken = await GitlabAppService.findByUserId(Types.ObjectId(id));
```
Comment (rekpero) [inline src/components/GitlabApp/service.ts:165]: same here: `gitUserToken`

File: src/components/Project/index.ts [hunk 5]
```diff
@@ -230,6 +252,29 @@ export async function getBitbucketRepositories(
...
+      argoDecodedHeaderToken
+    );
+
+    const repositories = await BitbucketAppService.getWorkspaceRepositories(
+      deserializedToken.session_id,
+      req.params.workspaceSlug
+    );
+    res.status(200).json({
+      success: true,
+      repositories: repositories,
```
Comment (rekpero) [inline src/components/Project/index.ts:271]:
```suggestion
      repositories,
```

File: src/components/Project/index.ts [hunk 6]
```diff
@@ -251,7 +296,55 @@ export async function getGitlabRepositories(
...
+  req: Request,
+  res: Response
+): Promise<void> {
+  try {
+    const argoDecodedHeaderToken: any = await JWTTokenService.decodeToken(req);
+    const deserializedToken: any = await JWTTokenService.verifyToken(
+      argoDecodedHeaderToken
+    );
+
+    const repositories = await GitlabAppService.getNamespaces(
```
Comment (rekpero) [inline src/components/Project/index.ts:309]: rename it to `namespaces`

File: src/components/Project/index.ts [hunk 7]
```diff
@@ -251,7 +296,55 @@ export async function getGitlabRepositories(
...
+      deserializedToken.session_id
+    );
+
+    res.status(200).json({
+      success: true,
+      namespaces: repositories.data,
```
Comment (rekpero) [inline src/components/Project/index.ts:315]:
```suggestion
      namespaces,
```

File: src/components/Project/index.ts [hunk 8]
```diff
@@ -251,7 +296,55 @@ export async function getGitlabRepositories(
...
+    const deserializedToken: any = await JWTTokenService.verifyToken(
+      argoDecodedHeaderToken
+    );
+    const repositories = await GitlabAppService.getNamespaceRepositories(
+      deserializedToken.session_id,
+      req.params.namespaceId
+    );
+    res.status(200).json({
+      success: true,
+      repositories: repositories,
```
Comment (rekpero) [inline src/components/Project/index.ts:337]:
```suggestion
      repositories,
```

File: src/templates/user-org-invite/ninatestapp.pem [hunk 9]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (rekpero) [inline src/templates/user-org-invite/ninatestapp.pem:1]: what is this doing here, please remove this and put it in your gitignore

**Review by rekpero — CHANGES_REQUESTED** [review]
Please check the comments, and resolve them accordingly  @nikoleta-vukajlovic @dusanstanis

File: src/routes/ProjectRouter.ts [hunk 10]
```diff
@@ -10,9 +10,23 @@ router.get(
   ProjectComponent.getBitbucketRepositories
 );
 
+router.get("/bitbucket/workspaces", ProjectComponent.getBitbucketWorkspaces);
+
+router.get(
+  "/workspaces/:workspaceSlug",
```
Comment (rekpero) [inline src/routes/ProjectRouter.ts:16]: rename it to `/bitbucket/app/:appId/repositories`

File: src/routes/ProjectRouter.ts [hunk 11]
```diff
@@ -10,9 +10,23 @@ router.get(
...
+  ProjectComponent.getBitbucketWorkspaceRepositories
+);
+
 router.get("/gitlab/repositories", ProjectComponent.getGitlabRepositories);
 
-router.get("/github/repositories", ProjectComponent.GetUserRepositories);
+router.get("/gitlab/namespaces", ProjectComponent.getGitlabNamespaces);
+
+router.get(
+  "/namespaces/:namespaceId",
```
Comment (rekpero) [inline src/routes/ProjectRouter.ts:25]: rename it to `/gitlab/app/:appId/repositories`

**Review by rekpero — CHANGES_REQUESTED** [review]
More comments @nikoleta-vukajlovic

File: src/components/GitRepository/index.ts [hunk 12]
```diff
@@ -0,0 +1,375 @@
+import { Request, Response } from "express";
+import JWTTokenService from "../Session/service";
+import GithubAppService from "../GitHubApp/service";
+import BitbucketAppService from "../BitbucketApp/service";
+import GitlabAppService from "../GitlabApp/service";
+const axiosInstance = require("axios").default;
```
Comment (rekpero) [inline src/components/GitRepository/index.ts:6]: why do we need this, when `axios` is already imported

File: src/components/GitRepository/index.ts [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+  }
+}
+
+export async function getGithubOwners(req: Request, res: Response) {
+  const argoDecodedHeaderToken: any = await JWTTokenService.decodeToken(req);
+  const deserializedToken: any = await JWTTokenService.verifyToken(
+    argoDecodedHeaderToken
+  );
+  const id = Types.ObjectId(deserializedToken.session_id);
+  const getUserToken = await GithubAppService.findByUserId(id);
```
Comment (rekpero) [inline src/components/GitRepository/index.ts:82]:
```suggestion
  const userToken = await GithubAppService.findByUserId(id);
```

File: src/components/GitRepository/index.ts [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+    __dirname,
+    `../../templates/user-org-invite/${config.githubApp.PEM_FILE_NAME}`
+  );
+
+  privateKey = fs.readFileSync(fullPath, "utf8");
+}
+
+//-----------GitHub Routes-----------------
+export async function githubAppAuth(req: Request, res: Response) {
+  const getUserToken = await GithubAppService.findByUserId(
```
Comment (rekpero) [inline src/components/GitRepository/index.ts:34]:
```suggestion
  const userToken = await GithubAppService.findByUserId(
```

File: src/components/BitbucketApp/service.ts [hunk 15]
```diff
@@ -112,14 +114,67 @@ const BitbucketAppService: IBitbucketAppTokenService = {
...
+        workspace: workspace.slug,
       });
       for (const repo of repos.data.values) {
         repositories.push(repo);
       }
     }
     return repositories;
   },
+  async getWorkspaces(id: string): Promise<any> {
+    let getUserToken = await BitbucketAppService.findByUserId(
```
Comment (rekpero) [inline src/components/BitbucketApp/service.ts:126]:
```suggestion
    let userToken = await BitbucketAppService.findByUserId(
```

File: src/components/BitbucketApp/service.ts [hunk 16]
```diff
@@ -112,14 +114,67 @@ const BitbucketAppService: IBitbucketAppTokenService = {
...
+    const bitbucket = new Bitbucket({
+      auth: { token: getUserToken.accessToken },
+    });
+    return await bitbucket.workspaces.getWorkspaces({});
+  },
+  async getWorkspaceRepositories(
+    id: string,
+    workspaceSlug: string
+  ): Promise<any> {
+    let getUserToken = await BitbucketAppService.findByUserId(
```
Comment (rekpero) [inline src/components/BitbucketApp/service.ts:142]:
```suggestion
    let userToken = await BitbucketAppService.findByUserId(
```

File: src/components/BitbucketApp/service.ts [hunk 17]
```diff
@@ -112,14 +114,67 @@ const BitbucketAppService: IBitbucketAppTokenService = {
...
+
+    const bitbucket = new Bitbucket({
+      auth: { token: getUserToken.accessToken },
+    });
+    return await bitbucket.repositories.list({
+      workspace: workspaceSlug,
+    });
+  },
+  async getBranches(id: string, branchesQuery: any): Promise<any> {
+    let getUserToken = await BitbucketAppService.findByUserId(
```
Comment (rekpero) [inline src/components/BitbucketApp/service.ts:157]:
```suggestion
    let userToken = await BitbucketAppService.findByUserId(
```

File: src/components/BitbucketApp/service.ts [hunk 18]
```diff
@@ -112,14 +114,67 @@ const BitbucketAppService: IBitbucketAppTokenService = {
...
+
+    const repo_slug = branchesQuery.split("/")[6];
+    const workspace = branchesQuery.split("/")[5];
+
+    const bitbucket = new Bitbucket({
+      auth: { token: getUserToken.accessToken },
+    });
+
+    const response = await bitbucket.repositories.listBranches({
+      repo_slug: repo_slug,
```
Comment (rekpero) [inline src/components/BitbucketApp/service.ts:172]:
```suggestion
      repo_slug,
```

File: src/components/BitbucketApp/service.ts [hunk 19]
```diff
@@ -112,14 +114,67 @@ const BitbucketAppService: IBitbucketAppTokenService = {
...
+      workspace: workspace,
```
Comment (rekpero) [inline src/components/BitbucketApp/service.ts:173]:
```suggestion
      workspace,
```

File: src/components/BitbucketApp/service.ts [hunk 20]
```diff
@@ -112,14 +114,67 @@ const BitbucketAppService: IBitbucketAppTokenService = {
...
+    return await bitbucket.repositories.list({
+      workspace: workspaceSlug,
+    });
+  },
+  async getBranches(id: string, branchesQuery: any): Promise<any> {
+    let getUserToken = await BitbucketAppService.findByUserId(
+      Types.ObjectId(id)
+    );
+    if (new Date() > getUserToken.expiresAt) {
+      getUserToken = await BitbucketAppService.refreshToken(Types.ObjectId(id));
```
Comment (rekpero) [inline src/components/BitbucketApp/service.ts:161]:
```suggestion
      userToken = await BitbucketAppService.refreshToken(Types.ObjectId(id));
```

File: src/components/GitRepository/index.ts [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+      branches: response.data.values,
+    });
+  } catch (error) {
+    throw new Error(error);
+  }
+}
+
+//-----------GitLab Routes-----------------
+export async function gitlabAppAuth(req: Request, res: Response) {
+  const getUserToken = await GitlabAppService.findByUserId(
```
Comment (rekpero) [inline src/components/GitRepository/index.ts:262]:
```suggestion
  const userToken = await GitlabAppService.findByUserId(
```

File: src/components/GitlabApp/service.ts [hunk 22]
```diff
@@ -79,16 +80,17 @@ const GitlabAppService: IGitlabAppTokenService = {
...
         accessToken: authData.data.access_token,
         refreshToken: authData.data.refresh_token,
         expiresAt: new Date(new Date().setHours(new Date().getHours() + 2)),
       },
       { new: true }
     );
-    return updated;
+    return updated[0];
   },
   async getRepositories(id: string): Promise<any> {
```
Comment (rekpero) [inline src/components/GitlabApp/service.ts:95]: do we need this function anymore?

Comment (rekpero) [reply src/components/GitlabApp/service.ts:95]: Same for github and bitbucket components too. I think the endpoint is removed so remove this function or comment it out and put a deprecated comment on top of it.

File: src/components/BitbucketApp/service.ts [hunk 23]
```diff
@@ -85,16 +86,17 @@ const BitbucketAppService: IBitbucketAppTokenService = {
...
+        {
+          accessToken: authData.data.access_token,
+          refreshToken: authData.data.refresh_token,
+          expiresAt: new Date(new Date().setHours(new Date().getHours() + 2)),
+        },
+        { new: true }
+      );
+    return updated[0];
   },
   async getRepositories(id: string): Promise<any> {
```
Comment (rekpero) [inline src/components/BitbucketApp/service.ts:101]: same here, do we need this func anymore?

**Review by rekpero — CHANGES_REQUESTED** [review]
@nikoleta-vukajlovic some more refactor needed in this PR, please resolve them

File: src/components/GitlabApp/service.ts (hunk 22 above)
Comment (nikoleta-vukajlovic) [reply src/components/GitlabApp/service.ts:95]: will remove

File: src/components/GitRepository/index.ts (hunk 12 above)
Comment (nikoleta-vukajlovic) [reply src/components/GitRepository/index.ts:6]: removed

**Pull request merged**